### PR TITLE
[superseded] js -d:nodejs now supports osenv: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`

### PR DESCRIPTION
**Pull request by timotheecour**
* nodejs now supports osenv: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`

see tests:
tests/js/tos.nim

* with js and not nodejs, we now show a proper compiler error msg instead of this crash:
```
osenv.nim(96, 17) Error: internal error: expr(nkBracketExpr, tyUncheckedArray)
        while gEnv[i] != nil:
```
this is the best we can do since js without nodejs can't have access to environment variables

* with `envPairs` in nimvm, we now show a cleaner error msg instead of
```
osenv.nim(272, 22) Error: cannot evaluate at compile time: environment
      for i in 0..high(environment):
```
but future PR's can make it work for nimvm, just as it does with `getEnv` + friends

### CODE REVIEW

**Comment by alaviss** [comment]
I thought the stdlib took a stance on not supporting nodejs?

**Comment by timotheecour** [comment]
there's already some support for nodejs. It doesn't add much complexity to support thanks to nodejs builtin libraries, and creates a lot of complexity in workarounds for client code that needs it, as any code that relies on getEnv + friends would have to use some custom `when defined(nodejs)` without this PR.

File: lib/pure/includes/osenv.nim [hunk 1]
```diff
@@ -3,110 +3,114 @@
...
-        if e == nil: return # an error occurred
-        while true:
-          var eend = strEnd(e)
-          add(environment, $e)
-          e = cast[WideCString](cast[ByteAddress](eend)+2)
-          if eend[1].int == 0: break
-        discard freeEnvironmentStringsW(env)
+when defined(js):
+  when not defined(nodejs):
+    {.error: "requires -d:nodejs".}
```
Comment (Araq) [inline lib/pure/includes/osenv.nim]: What if I compile to JS but use ``getEnv`` at compile-time in the VM?

**Comment by Araq** [comment]
I don't see how support 4 operations out of N helps Node.JS users. It's much easier to communicate that we simply don't support NodeJS in the standard library and it was only added for our own testing purposes.

**Comment by dom96** [comment]
Unless the additions to the NodeJS target help our tests then they shouldn't be accepted. I don't believe this helps our testing so it's time to close it.

File: lib/pure/includes/osenv.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/includes/osenv.nim]: I've fixed this in the latest commit, but this PR needs to be re-opened to view that commit

**Comment by timotheecour** [comment]
@araq can we please re-open? I've fixed https://github.com/nim-lang/Nim/pull/15475#discussion_r499271853 in latest commit which can't appear here because this PR was closed.

> I don't see how support 4 operations out of N helps Node.JS users. It's much easier to communicate that we simply don't support NodeJS in the standard library and it was only added for our own testing purposes.

I don't understand this stance against nodejs, it should IMO be supported on a best effort basis by accepting valid PRs to improve its situation. It's not like it's adding a lots of complexity, given the existing nodejs libraries: this PR is small and not a maintenance burden.
Relegating this to a nimble library introduces a lots of un-necessary complexity.

**Comment by disruptek** [comment]
Agree that should be supported and willing to write an RFC if necessary to change the expectation; I couldn’t find any argument there against node when I went looking...

**Comment by Araq** [comment]
I've reopened it in order not to choke off the discussion but the PR itself remains unacceptable. 

I remember going through os.nim to see which procs are supported by Nimscript, it costs effort to ensure it works well or either produces a good error message, node.js support will face the same problems. What's so bad about a separate **module** that becomes a node.js specific import statement? We should learn from our past mistakes, not repeat them.

**Comment by disruptek** [comment]
NimScript is not Node.  Do you expect to select your backend via an import statement?  I don't expect to make a separate `import nodejs` just to be able to use procedures whose API is expected to match the documentation for `import os`.

If you have a different idea about how this should work, please share it.  So far, I think the effort to support environment is very much worthwhile to include, because it makes code I have that runs on the cloud quite a bit easier to integrate.

**Comment by timotheecour** [comment]
Also, a separate `import nodejs` would fly in the face of reusability and would pretty much guarantee that code that works for `nim c` would not work for `nim js -d:nodejs` without littering it with `when defined(nodejs)` conditionals, which is impossible for 3rd party code.

The point of `std/os` + friends is to **abstract away platform/backend differences** as much as possible, so that client code can use it to build on top of it without re-inventing the wheel. `import nodejs` would make this impossible. This is why vm code (or js code) should look like regular nim code as much as possible, otherwise we're back to non portable code and software fragmentation.

File: lib/pure/includes/osenv.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/includes/osenv.nim]: see latest commit: `getEnv` now works in nim vm (with same semantics as before this PR, ie it returns env as known at CT)

**Comment by timotheecour** [comment]
PTAL

**Comment by Araq** [comment]
Don't "PTAL" me, I have already said, it's to enable a discussion. Your arguments are good, but this needs an RFC:

- Why we should support node.js in Nim's core libs.
- Why the tests we (will?) have suffice and produce good error messages when proc os.foo is not supported on node.js
- How to restructure the code in ``os.nim`` to have fewer ``when`` statements. (Probably via more ``include`` statements). Or more broadly: How to keep maintenance costs low so that we can support it for the next decade. Most costs in software development come **after** the code was written.

**Comment by Araq** [comment]
> Also, a separate import nodejs would fly in the face of reusability and would pretty much guarantee that code that works for nim c would not work for nim js -d:nodejs without littering it with when defined(nodejs) conditionals, which is impossible for 3rd party code.

Compare that to today's situation: We don't support node.js. If you need it, you need to write custom code. Not ideal for you, but much friendlier for 3rd party code and the ecosystem. Less effort spent on testing and PRs. Most 3rd party code won't work with node.js anyway, I couldn't care less if my libFFI or UI library works with node.js.

> The point of std/os + friends is to abstract away platform/backend differences as much as possible, so that client code can use it to build on top of it without re-inventing the wheel. import nodejs would make this impossible. This is why vm code (or js code) should look like regular nim code as much as possible, otherwise we're back to non portable code and software fragmentation.

Yes, but supporting a new platform is expensive, we don't support .NET or the JVM either. Not because it wouldn't be useful to somebody, but because we cannot. Haxe supports many more platforms than Nim and IMHO it's hurting Haxe, every library has to be tested against a plethora of targets and most people do not care about all the targets (in 2020 few care about PHP as a distribution platform or Flash).

**Comment by al6x** [comment]
Yea, agree with Araq. 

While I use node.js heavily, and would like to see a proper support for node.js. That would tax the Nim development.

The solution with using node.js imports sounds not ideal but more or less reasonable compromise.

Also, there's the node.js successor "deno", and it has different API than "node". Requiring Nim core to support all possible cimpilation targets like "node", "deno" etc. feels wrong and is not scalable.

Would be nice to support specific target via separate plugin (I don't know if that's possible), not tied to Nim's core. And use it like - at compile time you tell Nim compiler to replace nim std lib with the std lib from the plugin.

**Comment by disruptek** [comment]
I agree with @Araq as far as the assessment goes, but no one is asking to "support" node.  The goal is merely to not put technical or political obstacles in place, such as a divergent `import nodejs` or a platform like:

_I support the web-browser.  Also, I use Node to test web-browser code because I'm lazy.  So, like, submit any PR you want as long as it improves my silly use of Node to test web-browser code.  But even if your contribution is free and improves support for your usage of Nim, as well as attracting new users and enhancing platform support and testing, I'm not interested if it doesn't help me test a different platform with code that you're not interested in._

I mean, this is pretty off-putting, right?

**Comment by nc-x** [comment]
> The goal is merely to not put technical or political obstacles in place, such as a divergent import nodejs
  
tbh, I would like it if there was official support in the stdlib, it would make it easier to make the same code work on node as well as natively.
  
But...

> no one is asking to "support" node
  
How would this work? If people start using these functions in their code, they will post any bugs/issues they find. If people are using these functions, they will also wonder why only these are supported and not the stuff that they want to use.
  
So, somebody would have to maintain / "support" this stuff properly, no?
  
And if you look at the js backend, even that is just passively maintained, in fact, it has only seen major work done by two people - recloser and hlaaftana in a very long time IIRC. 
  
So, adding "support" for node would just add more code that people would be using but nobody would be maintaining.

**Comment by Araq** [comment]
Write an RFC and see how many upvotes it gets and then prepare the infrastructure to test os.nim's features for NodeJS. Things covered by CI don't bitrot as easily and the JS backend is still maintained even though the innovations happen for the C backend.

> I mean, this is pretty off-putting, right?

You can phrase everything as "off-putting": _"Here we added nodeJS support for os.nim but only for these 3 procs that Timothee needed. Now other contributors who seek to improve os.nim only need to read more 'when' statements and need to be careful not to break these 3 procs for Node.Js. Now aren't you happy with that."_

**Comment by timotheecour** [comment]
relevant: VSCode requires extensions be written in typescript, but can be written in `nim` thanks to `nim js -d:nodejs`.
This was in fact done recently, see https://github.com/saem/vscode-nim/blob/master/nimvscode.nimble which rewrote the code supporting nimscode-nim from typescript to nim.

Better nodejs support would simplify writing such code, enabling more code reuse from libraries not originally written with nodejs in mind (eg for https://github.com/saem/vscode-nim/blob/2cfb5face2b31a177ebdf9742a8a0913c1807194/src/nimvscode/nimUtils.nim), as opposed to code silos.

More generally, nodejs is a pretty popular language for writing plugins.

refs https://forum.nim-lang.org/t/7014#44124

**Comment by Araq** [comment]
> Better nodejs support would simplify writing such code, enabling more code reuse from libraries not originally written with nodejs in mind (eg for https://github.com/saem/vscode-nim/blob/2cfb5face2b31a177ebdf9742a8a0913c1807194/src/nimvscode/nimUtils.nim), as opposed to code silos.

Yet the author of vscode-nim didn't complain and accomplished the task successfully. 

You're moving the costs from the application developers to the Nim core developers. And this is not even a simple "move", writing library code is easily at least twice as expensive. When we decide to support node.js in Nim's standard library, soon enough we end up having to **emuate** osproc's quirks on top of node.js's process handling own quirks... And in the end somebody spent weeks on an emulation layer just so that somebody can simplify some application code. But application code is **much** simpler to write than library code!

And even if you want to have more library support for node.js, so write Nim libraries that support node.js, there is still no good justification for patching Nim's standard libraries.

**Comment by timotheecour** [comment]
All these arguments would apply just as well for the other backends: js, VM, nimscript or even posix vs windows.
But in the end the approach we used works best, we have a single os.nim + friends that abstracts platform differences and works with vm, posix, windows, js, with partial support and caveats implied for each backend, instead of having incompatible os_windows.nim, os_nimscript.nim etc, a ton of duplicated code and zero opportunities for code reuse.

> moving the costs from the application developers to the Nim core developers

s/core developers/Nim contributors/. I'm not asking that Nim should say it officially supports nodejs, I'm saying that Nim should have best-effort support for nodejs by welcoming PR's, and such PR's should strive to keep maintenance costs proportionally low.

> But application code is much simpler to write than library code!

Yes, because the scope is known, but it's still much easier to write when good library code is available. 

So the question becomes: how do we improve support for nodejs such that:
* we offer the same interface (no silos, `import os` should work)
* we minimizing impact on code health / maintenance.

One option is to write a separate `nodejs_os.nim` specific to `nodejs` (with most of the nodejs specific code) and adding some logic in os.nim to rexport it, taking care of minimizing duplication (eg for higher level procs that shouldn't be duplicated). EDIT: see https://github.com/nim-lang/Nim/pull/15826

**Comment by timotheecour** [comment]
closing for now in favor of https://github.com/nim-lang/Nim/pull/15826 which results in a better separation between nodejs code and regular code, which should be easier to maintain

**Pull request closed without merging**